File: src/tools/deploy-vault-cluster.ts
import { exec } from "child_process";
import { promisify } from "util";
import * as path from "path";

const execAsync = promisify(exec);

interface DeployVaultClusterArgs {
  configFiles: string[];
  deploymentType: "complete" | "unsealer-only" | "data-vault";
  skipCertificates?: boolean;
  skipPrerequisites?: boolean;
  workingDirectory?: string;
}

export async function deployVaultCluster(args: any) {
  const {
    configFiles,
    deploymentType,
    skipCertificates = false,
    skipPrerequisites = false,
    workingDirectory = process.cwd(),
  } = args as DeployVaultClusterArgs;

  try {
    // Validate config files exist
    const configFilePaths = configFiles.map((f) => path.join(workingDirectory, f));
    
    // Build the deployment command
    let command = `cd "${workingDirectory}" && ./setup-vault-cluster.sh`;
    
    // Add configuration files
    configFilePaths.forEach((configPath) => {
      command += ` ${path.basename(configPath)}`;
    });
    
    // Add flags
    if (skipCertificates) {
      command += " --skip-certificates";
    }
    if (skipPrerequisites) {
      command += " --skip-prerequisites";
    }

    // Execute deployment
    const { stdout, stderr } = await execAsync(command, {
      cwd: workingDirectory,
      maxBuffer: 1024 * 1024 * 10, // 10MB buffer for large outputs
    });

    let response = `# Vault Cluster Deployment\n\n`;
    response += `**Deployment Type:** ${deploymentType}\n`;
    response += `**Configuration Files:** ${configFiles.join(", ")}\n\n`;
    response += `## Deployment Output\n\n\`\`\`\n${stdout}\`\`\`\n`;
    
    if (stderr) {
      response += `\n## Warnings/Errors\n\n\`\`\`\n${stderr}\`\`\`\n`;
    }

    response += `\n## Next Steps\n\n`;
    if (deploymentType === "unsealer-only" || deploymentType === "complete") {
      response += `1. Initialize the unsealer vault: Use \`initialize_vault\` tool\n`;
      response += `2. Enable transit secrets engine\n`;
    }
    if (deploymentType === "data-vault" || deploymentType === "complete") {
      response += `3. Verify data vault auto-unsealed\n`;
      response += `4. Check cluster status with \`check_cluster_status\` tool\n`;
    }

    return {
      content: [
        {
          type: "text",
          text: response,
        },
      ],
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      content: [
        {
          type: "text",
          text: `Deployment failed: ${errorMessage}\n\nUse the troubleshoot_deployment tool to diagnose issues.`,
        },
      ],
      isError: true,
    };
  }
}

File: python_src/tools/deploy_vault_cluster.py
from __future__ import annotations

from pathlib import Path
from typing import Any
import shlex

from .common import text_response, run_shell, ToolResponse


def deploy_vault_cluster(args: dict[str, Any]) -> ToolResponse:
    config_files = args.get("configFiles", [])
    deployment_type = args.get("deploymentType")
    skip_certificates = bool(args.get("skipCertificates", False))
    skip_prerequisites = bool(args.get("skipPrerequisites", False))
    working_directory = args.get("workingDirectory") or str(Path.cwd())

    try:
        config_file_paths = [str(Path(working_directory) / f) for f in config_files]

        command = f"cd {shlex.quote(working_directory)} && ./setup-vault-cluster.sh"
        for config_path in config_file_paths:
            command += f" {shlex.quote(Path(config_path).name)}"

        if skip_certificates:
            command += " --skip-certificates"
        if skip_prerequisites:
            command += " --skip-prerequisites"

        stdout, stderr = run_shell(command, cwd=working_directory, max_bytes=10 * 1024 * 1024)

        response = "# Vault Cluster Deployment\n\n"
        response += f"**Deployment Type:** {deployment_type}\n"
        response += f"**Configuration Files:** {', '.join(config_files)}\n\n"
        response += f"## Deployment Output\n\n```\n{stdout}```\n"

        if stderr:
            response += f"\n## Warnings/Errors\n\n```\n{stderr}```\n"

        response += "\n## Next Steps\n\n"
        if deployment_type in ("unsealer-only", "complete"):
            response += "1. Initialize the unsealer vault: Use `initialize_vault` tool\n"
            response += "2. Enable transit secrets engine\n"
        if deployment_type in ("data-vault", "complete"):
            response += "3. Verify data vault auto-unsealed\n"
            response += "4. Check cluster status with `check_cluster_status` tool\n"

        return text_response(response)
    except Exception as error:  # noqa: BLE001
        return text_response(
            f"Deployment failed: {error}\n\nUse the troubleshoot_deployment tool to diagnose issues.",
            is_error=True,
        )
